zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/FastLeaderElection.java

```diff
@@ -977,6 +977,20 @@ public Vote lookForLeader() throws InterruptedException {
                     int tmpTimeOut = notTimeout * 2;
                     notTimeout = Math.min(tmpTimeOut, maxNotificationInterval);
                     LOG.info("Notification time out: {}", notTimeout);
+
+                    /*
+                     * When a leader failure happens on a master, the backup will be supposed to receive the honour from
+                     * Oracle and become a leader, but the honour is likely to be delay. We do a re-check once timeout happens
+                     *
+                     * The leader election algorithm does not provide the ability of electing a leader from a single instance
+                     * which is in a configuration of 2 instances.
+                     * */
+                    if (voteSet != null && voteSet.hasAllQuorums() && notTimeout != minNotificationInterval) {
```

Comment:
marking this as a scary part that i need to think about some more. (i haven't noticed anything wrong, but i want to come back to this.)